x/did/keeper/did.go

```diff
@@ -1,63 +1,90 @@
 package keeper
 
 import (
+	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	"github.com/allinbits/cosmos-cash/x/did/types"
 )
 
 func (k Keeper) SetDidDocument(ctx sdk.Context, key []byte, document types.DidDocument) {
-	k.Set(ctx, key, types.DidDocumentKey, document, k.MarshalDidDocument)
+	k.Set(ctx, key, types.DidDocumentKey, document, k.Marshal)
 }
 
 func (k Keeper) GetDidDocument(ctx sdk.Context, key []byte) (types.DidDocument, bool) {
 	val, found := k.Get(ctx, key, types.DidDocumentKey, k.UnmarshalDidDocument)
 	return val.(types.DidDocument), found
 }
 
+// UnmarshalDidDocument unmarshall a did document= and check if it is empty
+// ad DID document is empty if contains no context
 func (k Keeper) UnmarshalDidDocument(value []byte) (interface{}, bool) {
-	document := types.DidDocument{}
-
-	err := k.cdc.UnmarshalBinaryBare(value, &document)
-	if err != nil {
-		return types.DidDocument{}, false
-	}
+	data := types.DidDocument{}
+	k.Unmarshal(value, &data)
+	return data, types.IsValidDIDDocument(&data)
+}
 
-	if len(document.Context) == 0 {
-		return types.DidDocument{}, false
-	}
+func (k Keeper) SetDidMetadata(ctx sdk.Context, key []byte, meta types.DidMetadata) {
+	k.Set(ctx, key, types.DidMetadataKey, meta, k.Marshal)
+}
 
-	return document, true
+func (k Keeper) GetDidMetadata(ctx sdk.Context, key []byte) (types.DidMetadata, bool) {
+	val, found := k.Get(ctx, key, types.DidMetadataKey, k.UnmarshalDidMetadata)
+	return val.(types.DidMetadata), found
 }
 
-func (k Keeper) MarshalDidDocument(value interface{}) []byte {
-	did := value.(types.DidDocument)
+func (k Keeper) UnmarshalDidMetadata(value []byte) (interface{}, bool) {
+	data := types.DidMetadata{}
+	k.Unmarshal(value, &data)
+	return data, types.IsValidDIDMetadata(&data)
+}
 
-	bytes, _ := k.cdc.MarshalBinaryBare(&did)
+func (k Keeper) Marshal(value interface{}) (bytes []byte) {
+	switch value := value.(type) {
+	case types.DidDocument:
```

Comment:
If I am not mistaken you can use something on the style of case types.DidDocument, types.DidMetadata: and you can avoid repeating the encoding line.

Reply:
that is automatically casting `value` to the type in the  `case`,  with multiple case options the automatic casting will not work and value will be just an `interface{}`